pvlib/spectrum/mismatch.py

```diff
@@ -1100,3 +1101,95 @@ def qe_to_sr(qe, wavelength=None, normalize=False):
         spectral_responsivity = normalize_max2one(spectral_responsivity)
 
     return spectral_responsivity
+
+
+def average_photon_energy(spectral_irr):
+    r"""
+    Calculate the average photon energy of one or more spectral irradiance
+    distributions.
+
+    Parameters
+    ----------
+    spectral_irr : pandas.Series or pandas.DataFrame
+
+        Spectral irradiance, must be positive. [Wm⁻²nm⁻¹]
+
+        A single spectrum must be a :py:class:`pandas.Series` with wavelength
+        [nm] as the index, while multiple spectra must be a
+        :py:class:`pandas.DataFrame` with column headers as wavelength [nm].
+
+    Returns
+    -------
+    ape : numeric or array
+        Average Photon Energy [eV].
+
+    Notes
+    -----
+    The average photon energy (APE) is an index used to characterise the solar
+    spectrum. It has been used widely in the Physics literature since the
+    1900s, but its application for solar spectral irradiance characterisation
+    in the context of PV performance modelling was proposed in [1]_. The APE
+    is calculated based on the principle that a photon's wavelength is
+    inversely proportional to its energy:
+
+    .. math::
+
+        E_\gamma = \frac{hc}{\lambda},
+
+    where :math:`E_\gamma` is the energy of a photon with wavelength
+    :math:`\lambda`, :math:`h` is the Planck constant, and :math:`c` is the
+    speed of light. Therefore, the average energy of all photons within a
+    single spectral irradiance distribution provides an indication of the
+    general shape of the spectrum. A higher average photon energy
+    (shorter wavelength) indicates a blue-shifted spectrum, while a lower
+    average photon energy (longer wavelength) would indicate a red-shifted
+    spectrum. This value of the average photon energy can be calculated by
+    dividing the total number of photons in the spectrum by the total energy in
+    the spectrum as follows [1]_:
+
+    .. math::
+
+        \varphi = \frac{1}{q} \cdot \frac{\int_a^b \Phi_\lambda \, d\lambda}
+        {\int_a^b E_\lambda \, d\lambda}.
```

Comment:
Seems like the numerator and denominator need to be swapped in this equation.  Also, do we need to define $E_\lambda$?  Also, I initially misread it for the previously defined $E_\gamma$.  Maybe worth using a different letter for one of them.

Reply:
A proposal using alternative symbols: 

``` math
\overline{E_\gamma} = \frac{1}{q} \frac{\int G(\lambda)\,\mathrm{d}\lambda}{\int \Phi(\lambda)\,\mathrm{d}\lambda}
```

I think $\Phi(\lambda)$ should be defined as well.  I think including the integration limits may be a distraction, so I omitted them here. 

Should the `1/q` factor be omitted in this docs equation, as it is merely a unit conversion?

Reply:
+1 for your symbols suggestion.
 (separate issue/PR) Should symbols like $G(\lambda)$ and $\Phi_\lambda$ be added to the user guide variables and symbols page? So far there are only text/word variable name definitions there but I see benefit in defining these terms somewhere central.
 
 > Should the 1/q factor be omitted in this docs equation, as it is merely a unit conversion?

The physicist in me wants to nix the `1/q` for the same reason, but ultimately I think it is best to retain it. Since we are defining $\overline{E_\lambda}$ in eV, as is the convention in PV modelling, I think it would be easier for users to follow if we include the `1/q`. In addition, at least to my knowledge anyway, this is how the equation is usually written in other publications so there would be some consistency there (again, just easier for the user, especially if they're not super familiar with this metric)

I'll make the symbol changes and retain the `1/q` for now, but I'll leave this comment open for a bit in case anyone else wants to weigh in.

Reply:
Ok!  Perhaps the text around the equation could explain the purpose of the `1/q` factor.

Reply:
> Ok! Perhaps the text around the equation could explain the purpose of the `1/q` factor.

Current docstring (haven't changed symbols yet) says:

```
math:`q` is the elementary charge used here so that the
    average photon energy, :math:`\varphi`, is expressed in electronvolts (eV).
```

Do you mean this should be phrased in terms of `1/q` somehow rather than just explaining `q`?

Reply:
Ah sorry, I missed that it was already explained.  I retract my request for further discussion of `q` (or `1/q`).  Thanks!